### PR TITLE
ISSUE-638 Add APISIX in the demo OIDC docker compose on TMail

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/tmail-backend/issues/638

Ref: 
- How to configure run with standalone mode: https://github.com/apache/apisix-docker/blob/821b4744a003cdff5eff75c2a1896ec0eaa1089a/debian/docker-entrypoint.sh#L24

### CODE REVIEW

**Comment by vttranlina** [comment]
## Remark:
- Document for stand-alone mode not available from Apisix 3.2 (In 3.1 -> https://apisix.apache.org/docs/apisix/3.1/stand-alone/)

File: demo/docker-compose.yml [hunk 1]
```diff
@@ -101,5 +101,20 @@ services:
     networks:
       - tmail
 
+  apisix:
+    command: >
+      bash -c "cat /apisix-config/config.yaml > /usr/local/apisix/conf/config.yaml
+      && ln -s /apisix-config/apisix.yaml /usr/local/apisix/conf/apisix.yaml 
+      && /docker-entrypoint.sh docker-start"
```
Comment (quantranhong1999) [inline demo/docker-compose.yml:108]: Question: We can not mount directly these two files? (APISIX override them on startup?)

**Review by chibenwa — COMMENTED** [review]
Read it.

File: demo/apisix/conf/config.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+deployment:
+  role: data_plane
+  role_data_plane:
+    config_provider: yaml
```
Comment (quantranhong1999) [inline demo/apisix/conf/config.yaml:4]: Could we try putting the environment variable `APISIX_STAND_ALONE=true` for APISIX container?

I ran APISIX in standalone mode using command
`docker run -it --rm -e APISIX_STAND_ALONE=true apache/apisix:3.2.0-debian` without this conf file.

File: demo/apisix/conf/apisix.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+routes:
+  -
+    id: httpbin    #For test, TODO remove
```
Comment (quantranhong1999) [inline demo/apisix/conf/apisix.yaml:3]: Question: Does the hyphen need to stand alone in 1 line? Looks weird to me.

File: demo/docker-compose.yml (hunk 1 above)
Comment (vttranlina) [reply demo/docker-compose.yml:108]: I just re-tested, the mount volume work.
I will update commit 

// 
TLDR: When I set `APISIX_STAND_ALONE=true`, I got some errors related to the mounted file, permission, override...
So I manually set the value file in this way.
But now, we don't need env `APISIX_STAND_ALONE` :D

File: demo/apisix/conf/config.yaml (hunk 2 above)
Comment (vttranlina) [reply demo/apisix/conf/config.yaml:4]: As my comment: https://github.com/linagora/tmail-backend/pull/652#discussion_r1171125022

we can see the `docker-entrypoint.sh`, The `cat` command has been making "empty" file after the mount file

File: demo/apisix/conf/apisix.yaml (hunk 3 above)
Comment (vttranlina) [reply demo/apisix/conf/apisix.yaml:3]: I copied the format in the document example 
https://apisix.apache.org/docs/apisix/3.1/stand-alone/

Comment (vttranlina) [reply demo/apisix/conf/apisix.yaml:3]: Just test 
It works with bellow format 
```yaml
routes:
  - id: test   
    uri: /ping
```

I think it is just a format convention, no big deal, we can keep the format similar to apisix document

**Comment by vttranlina** [comment]
> ## Remark:
> * Document for stand-alone mode not available from Apisix 3.2 (In 3.1 -> https://apisix.apache.org/docs/apisix/3.1/stand-alone/)

Stand-Alone mode is still available, Document for that has been merged in the deployment page 
https://apisix.apache.org/docs/apisix/deployment-modes/#standalone

**Review by quantranhong1999 — APPROVED** [review]

File: demo/apisix/conf/config.yaml (hunk 2 above)
Comment (Arsnael) [reply demo/apisix/conf/config.yaml:4]: Hmmmm actually I gave it a quick look and it is quite confusing... I start to understand why @vttranlina is struggling with it.

Official doc website of APISIX: - https://apisix.apache.org/docs/apisix/deployment-modes/#standalone

Says pretty much add this piece of code in config.yaml to start in standalone mode (what Tung did)

The github repo of the docker apisix project: https://github.com/apache/apisix-docker#how-to-run-apisix-in-stand-alone-mode

Says to set up the env variale like @quantranhong1999 is suggesting. Notice that the link at the end of the section telling you you can know more details about standalone mode towards a doc page on the official website gets you to... a 404 :)

Something clearly not consistent here.

Digging a bit more though: https://github.com/apache/apisix-docker/issues/401

That I believe is the issue you mentionned somewhere @vttranlina ? Notice it has been closed because of inactivity, not because it maybe got solved... Do you still have that issue?

My guess is doc on github repo potentially outdated, and should follow the official one?

Comment (Arsnael) [reply demo/apisix/conf/config.yaml:4]: It seems adding the env variable is overriding the conf files with default standalone conf... from what I understood.

**Comment by Arsnael** [comment]
Is it working as standalone now then or is there still an issue @vttranlina ? :)

File: demo/apisix/conf/config.yaml (hunk 2 above)
Comment (vttranlina) [reply demo/apisix/conf/config.yaml:4]: > It seems adding the env variable is overriding the conf files with default standalone conf... from what I understood.

Yes, 
This bellow is a shell script of Apisix docker-entrypoint.sh 

```
PREFIX=${APISIX_PREFIX:=/usr/local/apisix}

if [[ "$1" == "docker-start" ]]; then
    if [ "$APISIX_STAND_ALONE" = "true" ]; then
        cat > ${PREFIX}/conf/config.yaml << _EOC_
deployment:
  role: data_plane
  role_data_plane:
    config_provider: yaml
_EOC_

        cat > ${PREFIX}/conf/apisix.yaml << _EOC_
routes:
  -
#END
_EOC_
```

We can see the command 
```
        cat > ${PREFIX}/conf/apisix.yaml << _EOC_
routes:
  -
#END
```
It overrides conf file with "empty routes"

**Comment by vttranlina** [comment]
> Is it working as standalone now then or is there still an issue @vttranlina ? :)

It works as standalone mode

File: demo/apisix/conf/config.yaml (hunk 2 above)
Comment (Arsnael) [reply demo/apisix/conf/config.yaml:4]: Yep saw it as well :)

**Review by Arsnael — APPROVED** [review]
I tested it, seems working like a charm! Good job @vttranlina :)
